File: lib/effects/shadows.dart
import 'dart:ui';
import 'package:libmonet/contrast/contrast.dart';
import 'package:libmonet/core/hex_codes.dart';
import 'package:libmonet/effects/protection.dart';
import 'package:libmonet/util/with_opacity_neue.dart';

export 'package:libmonet/effects/protection.dart'
    show StackedShadowSpec, getStackedShadowSpec, convertRadiusToSigma;

/// Flutter-flavored view of a [StackedShadowSpec].
///
/// The math lives in `protection.dart` (`getStackedShadowSpec`); this type
/// exists to hand `TextStyle.shadows` a ready list. Prefer the single-layer
/// halo (`getHalo` + a dilate-then-blur underlay) wherever the renderer
/// allows; stacked shadows are the compatibility fallback.
class ShadowResult {
  final double blurRadius;

  /// ARGB of the shadow color (typically black or white).
  final int shadowArgb;

  /// Opacities for each shadow layer. Identical by construction (closed
  /// form), unlike the legacy iterative solver's descending tail.
  final List<double> opacities;

  /// False when no stack of shadows can meet the target (infeasible
  /// protection, zero blur, or layer cap exceeded). The layers returned are
  /// then the best effort. Never silently absent.
  final bool meetsTarget;

  ShadowResult({
    required this.blurRadius,
    required this.shadowArgb,
    required this.opacities,
    this.meetsTarget = true,
  });

  List<Shadow>? _shadows;
  List<Shadow> get shadows {
    _shadows ??= opacities
        .map((e) => Shadow(
              color: Color(shadowArgb).withOpacityNeue(e),
              blurRadius: blurRadius,
              offset: const Offset(0, 0),
            ))
        .toList();
    return _shadows!;
  }
}

ShadowResult _fromSpec(StackedShadowSpec spec) => ShadowResult(
      blurRadius: spec.blurRadius,
      shadowArgb: spec.argb,
      opacities: spec.opacities,
      meetsTarget: spec.meetsTarget,
    );

/// Shadow layers so [foreground] meets [contrast] against every color in
/// [backgrounds].
ShadowResult getShadowOpacitiesForBackgrounds({
  required Color foreground,
  required Iterable<Color> backgrounds,
  required double contrast,
  required Algo algo,
  required double blurRadius,
  double contentRadius = -1.0,
  bool debug = false,
}) {
  return _fromSpec(getStackedShadowSpec(
    foregroundArgb: foreground.argb,
    backgroundArgbs: backgrounds.map((c) => c.argb).toList(),
    contrast: contrast,
    algo: algo,
    blurRadius: blurRadius,
    contentRadius: contentRadius,
  ));
}

File: js/src/shadows.ts
// Stacked-shadow wrappers over the exact protection solver (ADR-001 D-5).
// The math lives in protection.ts (getStackedShadowSpec); these exist for
// API compatibility. Direct port of lib/effects/shadows.dart.

import {type Argb} from './color.js';
import {Algo} from './contrast.js';
import {getStackedShadowSpec, type StackedShadowSpec} from './protection.js';

export {gauss1d, convertRadiusToSigma} from './protection.js';

export class ShadowResult {
  constructor(
    readonly blurRadius: number,
    /** ARGB of the shadow color (typically black or white). */
    readonly shadowArgb: Argb,
    /** Opacities for each shadow layer. Identical by construction. */
    readonly opacities: number[],
    /** False when no stack can meet the target (model-conditional; see StackedShadowSpec). */
    readonly meetsTarget: boolean = true,
  ) {}
}

function fromSpec(spec: StackedShadowSpec): ShadowResult {
  return new ShadowResult(spec.blurRadius, spec.argb, spec.opacities, spec.meetsTarget);
}

/**
 * Shadow layers so foreground meets contrast against every background — the
 * primary stacked-shadow entry point. All colors are checked exactly (no
 * min/max reduction; ADR-001 D-2).
 */
export function getShadowOpacitiesForBackgrounds(opts: {
  foreground: Argb; backgrounds: Iterable<Argb>; contrast: number; algo: Algo;
  blurRadius: number; contentRadius?: number;
}): ShadowResult {
  const backgroundArgbs = [...opts.backgrounds];
  return fromSpec(getStackedShadowSpec({
    foregroundArgb: opts.foreground,
    backgroundArgbs,
    contrast: opts.contrast,
    algo: opts.algo,
    blurRadius: opts.blurRadius,
    ...(opts.contentRadius === undefined ? {} : {contentRadius: opts.contentRadius}),
  }));
}

/** Shadow layers for a single known background. */
export function getShadowOpacitiesForColors(opts: {
  foreground: Argb; background: Argb; contrast: number; algo: Algo;
  blurRadius: number; contentRadius?: number;
}): ShadowResult {
  return getShadowOpacitiesForBackgrounds({...opts, backgrounds: [opts.background]});
}

